User Password reset

Shaggs opened the pull request:
# Description

User password reset using SMTP settings from pagermon. This will email the user a temporary password that can be used to login.

Fixes # (issue)

Added templates, JS and updated files

Please delete options that are not relevant.


- [X ] New feature (non-breaking change which adds functionality)


# How Has This Been Tested?

Tested on local device. New forgot password button added to logon.html


# Checklist:

- [X ] My code follows the style guidelines of this project
- [ X] I have performed a self-review of my own code
- [ X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated changelog.md with changes made

Commit:
changes to npm install for pm2

Commit:
extra fike not required

Shaggs committed:
Update README.md

Fix my edits

Commit:
Add PDRscaper & RPI installer

Shaggs committed:
Implement forgot password feature with SMTP support

Add SMTP configuration and temporary password generation for forgot password functionality.

Shaggs committed:
Password Rest

Added dependancies for pw reset

Shaggs committed:
Add user password reset script

This script allows for resetting user passwords by generating a temporary password and emailing it to the user.

Shaggs committed:
Add Forgot password functionality

Shaggs committed:
Add forgot password template

Shaggs committed:
Add 'Forgot password?' link to login template

eopo commented on client/pdr.py:
Please remove unrelated files from this PR

eopo commented on server/package.json:
```diff
@@ -56,7 +57,7 @@
     "slack": "latest",
     "socket.io": "^2.0.2",
-    "sqlite3": "~4.0.2",
+    "sqlite3": "^5.1.7",
```
Does it still run the tests? Some sqlite changes are breaking!

Shaggs replied:
I have just done a fresh setup and seems to be running with no issues

eopo replied:
Okay, tests via github actions currently fail because they were pulled from the master branch, which still tries to use node 12. We'll have to re-try the current test suite for a usable result.

eopo commented on server/routes/auth.js:
```diff
@@ -2,11 +2,14 @@ const express = require('express');
+const nodemailer = require('nodemailer');
 const confFile = './config/config.json';
-nconf.file({ file: confFile });
+const confDefault = './config/default.json';
```
Currently, if there is no config file, the default will be copied as config.json. Because of that i don't think there's a need to load the defaults again.
Actually, it is not necessary to re-setup nconf (giving it the config file etc.) per file, as it is already setup in `app.js`, which in turn requires all the other files - That makes them run in the same scope, so they have access to the already loaded nconf configuration. I don't blame you anyways, because our whole code does that too (We should remove that sometimes)

Shaggs replied:
Will look at this later today

eopo commented on server/routes/auth.js:
```diff
@@ -28,6 +31,47 @@ const lockoutCallback = function(req, res, next, nextValidRequestDate) {
+                .slice(0, 12);
+
+const loadSmtpConfig = () => {
+        const smtpSettings = nconf.get('plugins:SMTP');
```
I get your point here, but a core feature should not rely on plugin configuration. 
It would be better to add a new mail section to the core config and make the smtp-plugin use that.

Shaggs replied:
I did it this way so it would get the built in config rather than another file or hard coding SMTP settings. 

Thought was easier just to pull it all from one place

eopo replied:
Yeah, and you're right with that. I just would do it vice versa. Don't let the core functionality (Password Reset) use a Plugin's config options, but let the Plugin use the core config options instead.

In my opinion, the following would be the best
```json
{
    ... configOptions,
    "db": { ... }
    "mail": {
        ... smtp settings
    },
    "plugins": {
        "smtp" : { ... }
    }
    ...
}
```

The password reset only ever uses the main 'mail' settings. These settings could be used for future stuff relying on email, support for API based mail services could be added etc.
The smtp plugin uses the plugin configuration. If there is none, it uses the main mail settings. (So there's no need to configure twice, if you only plan to use only one setting).

I think relying on the plugin configuration is technical debt that could bite us in the ass in the future. If we ever went to move plugin configurations to a different format, e.g. move it to the database, we'd have to pull that apart.

eopo commented on server/routes/auth.js:
```diff
@@ -131,6 +175,63 @@ router.route('/logout').get(authHelper.isLoggedIn, function(req, res) {
+                if (!smtpConfig) {
+                        return res.status(500).send({
+                                status: 'failed',
+                                error: 'SMTP is not configured. Please set the SMTP plugin details in Admin > Settings.',
```
The route should be accessible to users, but gives a hint directed to the admin. Although I don't think that this is security relevant in this case, I'd prefer giving a more general hint (Password reset not possible, please contact the administrator). I would wish for a log entry as well, so the admin can check the logs for why the reset failed.

Shaggs replied:
I'll adjust the wording for this today. Play even go down the path of a checkbox in the plugin for user password reset

eopo commented on server/routes/auth.js:
```diff
@@ -131,6 +175,63 @@ router.route('/logout').get(authHelper.isLoggedIn, function(req, res) {
+
+                try {
+                        const user = await db('users')
+                                .whereRaw('LOWER(email) = LOWER(?)', [email])
```
`.whereRaw('LOWER(??) = LOWER(?)', 'email', email)` so it uses the right ticks for all database drivers

eopo commented on server/routes/auth.js:
```diff
@@ -131,6 +175,63 @@ router.route('/logout').get(authHelper.isLoggedIn, function(req, res) {
+                                .whereRaw('LOWER(email) = LOWER(?)', [email])
+                                .first();
+
+                        if (user) {
```
It's okay that way, but for legibility reasons I'd prefer an early exit like you did with missing smtp config:
```
if (!user) return ...
const tempPassword = ...
```

That removes one layer of indentation and for future edits, there's no need to check which if belongs to the else, because there's no else.

eopo commented on server/rpi-install.sh:
Please remove unrelated files from PR, you're welcome to submit seperate PRs for more features

eopo commented on server/scripts/user-password-reset.js:
Again, the whole project isn't a good example in that, but it from a DRY perspective, this is really redundant. I'd propose adding a `function resetPassword(email)` function, which is called by the router as well as the user-password-reset script. That way, you'd need no duplicated code and it would be easier to maintain the whole thing.

eopo reviewed (commented):
Looks really good to me. I've annotated a few things in the code.
I can't really say much for the frontend, because I really suck at frontend, so i've focussed on the backend code.

Shaggs committed:
Delete client/pdr.py

Shaggs committed:
Delete server/rpi-install.sh

Shaggs committed:
Delete server/scripts/user-password-reset.js

Removed as handled in routes.

Shaggs committed:
Remove loading of default configuration file.

Remove config file as not needed

Shaggs committed:
Change SMTP error message for password reset

Updated error message for self-service password reset when SMTP is not configured.

Shaggs reviewed (commented):
Updates done to code as per comments. All changes that have been made have been tested prior

Shaggs committed:
Merge branch 'develop' into password-reset
